cumulusci/util/xml/metadata_tree.py

```diff
@@ -0,0 +1,222 @@
+from typing import Union
+
+from lxml import etree
+
+from .salesforce_encoding import serialize_xml_for_salesforce
+
+METADATA_NAMESPACE = "http://soap.sforce.com/2006/04/metadata"
+
+
+def parse(source):
+    """Parse a file by path or file object into a Metadata Tree
+
+    The parse() function supports any of the following sources:
+
+        * an open file object (make sure to open it in binary mode)
+        * a file-like object that has a .read(byte_count) method returning a byte string on each call
+        * a filename string
+        * an HTTP or FTP URL string
+
+    Note that passing a filename or URL is usually faster than passing an open file or file-like object. However, the HTTP/FTP client in libxml2 is rather simple, so things like HTTP authentication require a dedicated URL request library, e.g. urllib2 or requests. These libraries usually provide a file-like object for the result that you can parse from while the response is streaming in.
+    """
+
+    doc = etree.parse(source)
+    return MetadataElement(doc.getroot())
+
+
+def fromstring(source):
+    """Parse a Metadata Tree from a string"""
+    return MetadataElement(etree.fromstring(source))
+
+
+class MetadataElement:
+    '''A class for representing Metadata in a Pythonic tree.
+
+    After you parse into a MetadataElement tree, you can refer
+    to child elements like this:
+
+    >>> from cumulusci.util.xml import metadata_tree
+    >>> Package.types.name.text
+    'StandardValueSet'
+    >>> Package = metadata_tree.fromstring(b"""<?xml version="1.0" encoding="UTF-8"?>
+    ... <Package xmlns="http://soap.sforce.com/2006/04/metadata">
+    ...     <types>
+    ...         <members>CaseReason</members>
+    ...         <members>CaseStatus</members>
+    ...         <members>CaseType</members>
+    ...         <name>StandardValueSet</name>
+    ...     </types>
+    ...     <version>46.0</version>
+    ... </Package>""")
+    >>> Package.types.name.text
+    'StandardValueSet'
+
+    Or you can refer to them like this:
+
+    >>> Package["types"]["name"].text
+    'StandardValueSet'
+
+    That might be convenient if you had the key names in a variable or if you had to refer to an element name
+    which would clash with a Python keyword (like `import`) or an instance method (like `text`).
+
+    You can refer to members of lists by index:
+
+    >>> Package["types"]["members"][1]
+    <members>CaseStatus</members> element
+
+    There are also methods for finding, appending, inserting and removing nodes, which have their own documentation.
+    '''
+
+    __slots__ = ["_element", "_parent", "_ns", "tag"]
+
+    def __init__(
+        self, element: etree._Element, parent: etree._Element = None,
+    ):
+        assert isinstance(element, etree._Element)
+        assert len(element.nsmap) == 1, "Only one namespace allowed"
+        self._element = element
+        self._parent = parent
```

Comment:
No, I think that it will break logic if the parent goes away before the child. In order to implement

foo.bar[3] we do foo.bar.parent.find['foo'][3]

Which I think answers another one of your questions.

We do it that way because we want foo.bar to behave as a regular leaf UNLESS we index it and then it becomes more like a list. So faking that means we need all of the information available to treat it as either a node or a list.

Reply:
By the way, this was the central trick that made any of it possible. In my first prototype I was parsing the schema to figure out what were "just" elements and what were lists. Then I realized I could use this trick to "do what the user expects" in either case.

Reply:
Ah yes, okay, this is the key trick that I wasn't quite spotting.

Reply:
It would really be worth a comment or docstring in `__getitem__`.

Reply:
e9deea920e2b8e2c